Created sensorCachingService and corresponding tests

larsverp opened the pull request:
The sensorCachingService keeps an object of all previous sensor measurements.

larsverp committed:
Created sensorCachingService and corresponding tests

The sensorCachingService keeps an object of all previous sensor measurements.

larsverp committed:
Merge branch 'develop' into feature/caching-service

BondarencoM commented on services/sensorCachingService.js:
```diff
@@ -0,0 +1,84 @@
+ * @author Lars
+ * @param {floor} floor The floor you want to recieve the max temp of.
+ */
+function getMaxTemp(floor){
+    var highestTemp
+    for(var element of sensorsFromFloor(floor)){
+        if(!highestTemp || element.value > highestTemp){
+            highestTemp = element.value
+        }
+    }
+    return highestTemp
+}
```
This compares temperatures to humidity because `value `could be either temperature or humidity and no check for `parameter` is made.

BondarencoM replied:
Btw, we do implement the policy that any bug needs to have a test proving that the bug was solved, right?

larsverp replied:
Created a `BUGFIX - highest temprature function should not return a humidity value` test function that's now working.

BondarencoM commented on services/sensorCachingService.js:
```diff
@@ -0,0 +1,84 @@
+ * @author Lars
+ * @param {floor} floor The floor you want to recieve the min temp of.
+ */
+function getMinTemp(floor){
+    var lowestTemp
+    for(var element of sensorsFromFloor(floor)){
+        if(!lowestTemp || element.value < lowestTemp){
+            lowestTemp = element.value
+        }
+    }
+    return lowestTemp
+}
```
Same as above

larsverp replied:
Created a `BUGFIX - lowest temprature function should not return a humidity value` test function that's now working.

BondarencoM commented on services/sensorCachingService.js:
```diff
@@ -0,0 +1,84 @@
+ * @author Lars
+ * @param {entry} entry The whole sensor entry
+ */
+function getKey(entry){
```
Since we're here, this name could be a bit more descriptive, something along the lines of `generateCacheKey` or `getEntryIdentifier` or something similar

BondarencoM commented on services/sensorCachingService.js:
```diff
@@ -0,0 +1,84 @@
+ * @param {entry} entry The whole sensor entry
+ */
+function getKey(entry){
+    if(!entry.loc_x || !entry.loc_y || !entry.floor || !entry.parameter || !entry.value){
+        throw new Error('1 or more fields are missing')
+    }
+    return entry.loc_x+'-'+entry.loc_y+'-'+entry.floor+'-'+entry.parameter
+}
```
```suggestion
    if(entry.loc_x && entry.loc_y && entry.floor && entry.parameter && entry.value){
         return `${entry.loc_x}-${entry.loc_y}-${entry.floor}-${entry.parameter}`   
    }
   throw new Error('1 or more fields are missing')
}
```

BondarencoM reviewed (dismissed):
Is great, requires a few small fixes.

larsverp committed:
Added new BUGFIX tests and made code work accordingly

Changed the way specific floor sensors are retrieved. The function name is changed and it will not longer return a humidity value

larsverp committed:
Merge branch 'feature/caching-service' of https://github.com/BondarencoM/KoziDataService into feature/caching-service

larsverp commented:
Wrote 2 BUGFIX tests that do **NOT** replace the old gethighest and getlowest tests. In this way the new bugfix tests will fail if the error ever returns.

BondarencoM commented on services/sensorCachingService.js:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Retrieve the maximum temprature/humidity for a specific floor.
+ * @author Lars
+ * @param {floor} floor The floor you want to recieve the max temp of.
```
here in the curly braces {} should be the data type of the parameter. Now I don't know what conventions are there but it seems reasonable to sue typescript's so that would be {number}
```suggestion
 * @param {number} floor The floor you want to recieve the max temp of.
```

BondarencoM reviewed (approved):
This is awesome ✨

BondarencoM commented on services/sensorCachingService.js:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Return array of all sensors on the given floor number
+ * @author Lars
+ * @param {floor} floor
```
```suggestion
 * @param {number} floor
```

oererenn reviewed (approved):
Nice Work
